File: nimbus/Views/DetailVC.m
//
//  DetailVC.m
//  nimbus
//
//  Created by Pericles Maravelakis on 19/04/2017.
//	periclesm@cloudfields.net
//	Licensed under Creative Commons Attribution-ShareAlike 4.0 International (CC BY-SA 4.0)
//	https://creativecommons.org/licenses/by-sa/4.0/
//

#import "DetailVC.h"
@import SafariServices;

@interface DetailVC ()
{
	IBOutlet UIImageView *clImage;
	IBOutlet UILabel *clInitials;
	IBOutlet UILabel *clName;
	IBOutlet UILabel *clAltitude;
	IBOutlet UILabel *clDetails;
}

@end

@implementation DetailVC

- (void)viewDidLoad
{
	[self GetCloudData];
    [super viewDidLoad];
}

#pragma mark - Data

- (void)GetCloudData
{	
	CloudList *cl = [Presenter GetCloudInfo:self.objectId];
	
	self.title = cl.name;
	clInitials.text = cl.initials;
	clName.text = cl.name;
	clAltitude.text = [[Presenter GetCloudType:cl.type] stringByAppendingString:@" altitude"];
	clDetails.text = [Presenter GetCloudDetails:cl.detail shortText:NO];

	clImage.alpha = 0;
	[Networker GetRemoteImage:[Presenter GetCloudImageURL:cl.detail] completion:^(UIImage *image) {
		self->clImage.image = image;
		[UIView animateWithDuration:0.25 animations:^{ self->clImage.alpha = 1; }];
	}];
}

#pragma mark - IBActions

- (IBAction)WikiButton:(id)sender
{
	CloudList *cl = [Presenter GetCloudInfo:self.objectId];
	SFSafariViewController *safariView = [[SFSafariViewController alloc] initWithURL:[NSURL URLWithString:[Presenter GetCloudWikiURL:cl.detail]]];
	[self presentViewController:safariView animated:YES completion:nil];
}

@end

File: nimbus/UIKit/DetailVC.swift
//
//  DetailVC.swift
//  nimbus
//
//  Created by Pericles Maravelakis on 1/2/20.
//  periclesm@cloudfields.net
//  Licensed under Creative Commons Attribution-ShareAlike 4.0 International (CC BY-SA 4.0)
//  https://creativecommons.org/licenses/by-sa/4.0/
//

import UIKit
import SafariServices

class DetailVC: UITableViewController {
    
    let vm = DetailVM()
    
    @IBOutlet weak var clImage: UIImageView!
    @IBOutlet weak var clInitials: UILabel!
    @IBOutlet weak var clName: UILabel!
    @IBOutlet weak var clAltitude: UILabel!
    @IBOutlet weak var clDetails: UILabel!
    @IBOutlet weak var activity: UIActivityIndicatorView!

    override func viewDidLoad() {
        super.viewDidLoad()
    }
	
	override func viewWillAppear(_ animated: Bool) {
		super.viewWillAppear(animated)
		self.getCloudData()
	}
    
    // MARK: - Data
    
    func getCloudData() {
		self.title = vm.cloud?.name
        clInitials.text = vm.cloud?.initials
        clName.text = vm.cloud?.name
		clAltitude.text = (vm.cloud?.type!.name)! + " altitude"
		clDetails.text = vm.cloud?.detail?.detail
		self.getCloudImage()
        
    }
	
	private func getCloudImage() {
		clImage.alpha = 0
		activity.startAnimating()
		
		Task(priority: .medium) {
			let imageURL = vm.getImageURL(vm.cloud?.objectId)
			self.clImage.image = await vm.asyncCloudImage(imageURL: imageURL)
			self.activity.stopAnimating()
			UIView.animate(withDuration: 0.25) {
				self.clImage.alpha = 1
			}
		}
	}

    // MARK: - Actions
    
    @IBAction func WikiButton(_ sender: UIBarButtonItem) {
		if let url = vm.getWikiURL(vm.cloudId) {
			let config = SFSafariViewController.Configuration()
			config.barCollapsingEnabled = true
			config.entersReaderIfAvailable = true
			
			let safari = SFSafariViewController(url: url, configuration: config)
			safari.modalPresentationStyle = .formSheet
			safari.modalTransitionStyle = .coverVertical
            safari.preferredControlTintColor = UIColor.systemIndigo
            
			self.present(safari, animated: true, completion: nil)
		}
	}
}
